base/shell/explorer/CTrayShowDesktopButton.cpp

```diff
@@ -0,0 +1,405 @@
+        
+        HWND taskbarWnd;
+        if (GetTaskbar(&taskbarWnd))
+            ::PostMessage(taskbarWnd, WM_NCPAINT, 0, 0);
+    }
+
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnDestroy(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (m_hTheme)
+    {
+        CloseThemeData(m_hTheme);
+        m_hTheme = NULL;
+    }
+    return 0;
+}
+
+VOID CTrayShowDesktopButton::OnDraw(HDC hdc, LPRECT prc)
+{
+    RECT rc = { prc->left, prc->top, prc->right, prc->bottom };
+    LPRECT lpRc = &rc;
+    HBRUSH hbrBackground = NULL;
+    if (m_hTheme)
+    {
+        HTHEME theme;
+        int part = 0;
+        int state = 0;
+        if (m_drawWithDedicatedBackground)
+        {
+            theme = m_hTheme;
+            if (m_bPressed)
+                state = PBS_PRESSED;
+            else if (m_bHovering)
+                state = PBS_HOT;
+            else
+                state = PBS_NORMAL;
+        }
+        else
+        {
+            part = TP_BUTTON;
+            theme = m_hFallbackTheme;
+
+            if (m_bPressed)
+                state = TS_PRESSED;
+            else if (m_bHovering)
+                state = TS_HOT;
+            else
+                state = TS_NORMAL;
+            
+            if (IsHorizontal)
+                rc.right -= m_inset.cx;
+            else
+                rc.bottom -= m_inset.cy;
+        }
+        
+        if (::IsThemeBackgroundPartiallyTransparent(theme, part, state))
+            ::DrawThemeParentBackground(m_hWnd, hdc, NULL);
+
+        ::DrawThemeBackground(theme, hdc, part, state, lpRc, lpRc);
+    }
+    else
+    {
+        hbrBackground = ::GetSysColorBrush(COLOR_3DFACE);
+        ::FillRect(hdc, lpRc, hbrBackground);
+
+        if (m_bPressed || m_bHovering)
+        {
+            UINT edge = m_bPressed ? BDR_SUNKENOUTER : BDR_RAISEDINNER;
+            DrawEdge(hdc, lpRc, edge, BF_RECT);
+        }
+    }
+
+    if (m_highContrastMode || !m_drawWithDedicatedBackground)
+    {
+        /* Prepare to draw icon */
+
+        int iconSize = 16;
+        // Used for icon centering further down
+        int iconHalfSize = iconSize / 2;
+        
+
+        // Determine X-position of icon's top-left corner
+        int iconX = rc.left;
+        int width = (rc.right - iconX);
+        iconX += (width / 2);
+        iconX -= iconHalfSize;
+
+        // Determine Y-position of icon's top-left corner
+        int iconY = rc.top;
+        int height = (rc.bottom - iconY);
+        iconY += (height / 2);
+        iconY -= iconHalfSize;
+
+        // Ok now actually draw the icon itself
+        if (m_icon)
+        {
+            DrawIconEx(hdc
+                , iconX
+                , iconY
+                , m_icon
+                , iconSize
+                , iconSize
+                , 0
+                , hbrBackground
+                , DI_NORMAL
+            );
+        }
+        else // Fallback for if icon isn't available or something idk lol
```

Comment:
@learn-more If memory serves, I implemented this back when I'd still planned to retrieve the icon from `imageres.dll`, due to Windows Server 2003 not having `imageres.dll`.

I'm not sure if it's actually needed, now that I'm using the icon from `shell32.dll` instead, since that one exists in ROS and every version of Windows *at least* since Server 2003...so uh, do you think I should keep it or get rid of it?